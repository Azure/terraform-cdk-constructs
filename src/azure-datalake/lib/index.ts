export * from "./filesystem";
export * from "./path";
export * from "./datalake";
