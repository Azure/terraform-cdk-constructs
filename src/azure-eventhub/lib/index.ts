export * from "./authorization";
export * from "./eventhubcluster";
export * from "./consumer";
export * from "./instance";
export * from "./kusto-connection";
export * from "./namespace";
