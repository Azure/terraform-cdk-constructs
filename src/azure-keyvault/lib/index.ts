export * from './certificate'
export * from './key'
export * from './policy'
export * from './secret'
export * from './vault'

