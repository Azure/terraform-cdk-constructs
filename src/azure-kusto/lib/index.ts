export * from "./cluster";
export * from "./compute-specification";
export * from "./database";
