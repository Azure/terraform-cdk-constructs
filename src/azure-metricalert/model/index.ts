export * from "./action-props";
export * from "./criteria-base-props";
export * from "./criteria-props";
export * from "./dynamic-criteria-props";
export * from "./metric-alert-props";
export * from "./criteria-dimension-props";
export * from "./function";
