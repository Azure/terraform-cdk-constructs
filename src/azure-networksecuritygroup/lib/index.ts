export * from './preconfigured-rules';
export * from './securitygroup';
