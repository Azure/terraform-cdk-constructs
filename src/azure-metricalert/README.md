# Azure Metric Alert Construct

This class represents a Metric Alert resource in Azure.

## What is Azure Metric Alert?

An Azure Metric Alert monitors a pre-computed and pre-aggregated metric over a period of time. We can use Azure provided metrics (see official doc for [dimensions supported](https://learn.microsoft.com/en-us/azure/azure-monitor/alerts/alerts-metric-near-real-time#metrics-and-dimensions-supported) ) or custom metrics. There are two types of Metric Alerts: `Static alert` and `Dynamic alert`. `Static Metric Alert` is given a static threshold to monitor, whereas `Dynamic Metric Alert` is leverage Azure Machine Learning to learn the normal pattern of metric and alert when the metric is outside of the normal pattern. The alert can also be configured to auto-mitigate when the metric returns to a healthy state.

## Azure Metric Alert Class Properties

This class has several properties that control the Alert Rules:

- `name` - The name of the Metric Alert.
- `resourceGroupName` - The name of the resource group in which the Metric Alert is created.
- `scopes` - A set of strings of resource IDs at which the metric criteria should be applied.
- `criteria` - (Optional) One ore more criteria. Either Criteria or dynamicCriteria is required.
- `dynamicCriteria` - (Optional) One ore more dynamic criteria. Either Criteria or dynamicCriteria is required.
- `enabled` - (Optional) Should this Metric Alert be enabled? Defaults to `true`.
- `automitigate` - (Optional) Should the alerts in this Metric Alert be auto resolved? Defaults to `true`.
- `frequency` - (Optional) The evaluation frequency of this Metric Alert, represented in ISO 8601 duration format. Possible values are PT1M, PT5M, PT15M, PT30M and PT1H. Defaults to `PT5M`.
- `windowSize` - (Optional) The period of time that is used to monitor alert activity, represented in ISO 8601 duration format. This value must be greater than frequency. Possible values are PT1M, PT5M, PT15M, PT30M, PT1H, PT6H, PT12H and P1D. Defaults to `PT5M`.
- `severity` - (Optional) The severity of this Metric Alert. Possible values are 0, 1, 2, 3 and 4. Defaults to `3`.
- `description` - (Optional) The description of this Metric Alert.
- `action` - (Optional) The action block of this Metric Alert.
- `targetResourceType` - (Optional) The resource type (e.g. Microsoft.Compute/virtualMachines) of the target resource. This is Required when using a Subscription as scope, a Resource Group as scope or Multiple Scopes.
- `targetResourceLocation` - (Optional) The location of the target resource. This is Required when using a Subscription as scope, a Resource Group as scope or Multiple Scopes.
- `tags` - (Optional) A mapping of tags to assign to the resource.

## Deploying a Metric Alert

You can deploy a Metric Alert using this class like so:

```typescript
  // Create a Resource Group first
  import * as rg from "../azure-resourcegroup";
  const resourceGroup = new rg.Group(this, "myResourceGroup", {
    name: 'myResourceGroup',
    location: 'eastus',
  });

  // Create a Log Analytics Workspace
  import * as law from "../azure-loganalytics";
  const logAnalyticsWorkspace = new la.Workspace(this, 'myLogAnalytics', {
      name: 'myLogAnalytics',
      location: 'eastus',
      resource_group_name: resourceGroup.name,
    });

  // Create a Metric Alert with defult settings in Log Analytics Workspace
  import * as ma from "./lib/azure-metricalert";
  new ma.MetricAlert(this, 'metricAlert', {
    name: `myMetricalert`,
    resourceGroupName: resourceGroup.name,
    scopes: [logAnalyticsWorkspace.id],
    criteria: [
      {
        metricName: "Heartbeat",
        metricNamespace: "Microsoft.operationalinsights/workspaces",
        aggregation: "Average",
        operator: "LessThan",
        threshold: 100,
      },
    ],
  });

```

Full example can be found [here](test/ExampleMetricAlert.ts).
