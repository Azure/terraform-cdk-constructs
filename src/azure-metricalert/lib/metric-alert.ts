import { MonitorMetricAlert } from "@cdktf/provider-azurerm/lib/monitor-metric-alert";
import * as cdktf from "cdktf";
import { Construct } from "constructs";
import * as moment from "moment";
import * as model from "../model";

export class MetricAlert extends Construct {
  readonly props: model.MetricAlertProps;
  public readonly id: string;

  /**
   * Constructs a new instance of the MetricAlert class.
   *
   * @param scope - The scope in which this construct is defined.
   * @param id - The ID of this construct.
   * @param props - The properties required for Metric Alert.
   */
  constructor(scope: Construct, id: string, props: model.MetricAlertProps) {
    super(scope, id);

    this.props = props;

    // Setup default values
    this.props.enabled = props.enabled ?? true;
    this.props.automitigate = props.automitigate ?? true;
    this.props.frequency = props.frequency ?? "PT5M";
    this.props.severity = props.severity ?? 3;
    this.props.windowSize = props.windowSize ?? props.frequency ?? "PT5M";

    // Properties validation
    this.ValidatePropsFrequency();
    this.ValidatePropsWindowSize();
    this.ValidatePropsWindowSizeGreaterThanFrequency();

    // Create Metric Alert
    const metricAlert = new MonitorMetricAlert(this, "metricAlert", {
      name: props.name,
      resourceGroupName: props.resourceGroupName,
      scopes: props.scopes,
      description: props.description,
      targetResourceType: props.targetResourceType,
      targetResourceLocation: props.targetResourceLocation,
      action: cdktf.listMapper(
        model.monitorMetricAlertActionToTerraform,
        true,
      )(this.props.action),
      tags: props.tags,
      enabled: this.props.enabled,
      autoMitigate: this.props.automitigate,
      frequency: this.props.frequency,
      severity: this.props.severity,
      windowSize: this.props.windowSize,
      criteria: cdktf.listMapper(
        model.monitorMetricAlertCriteriaToTerraform,
        true,
      )(this.props.criteria),
      dynamicCriteria: cdktf.listMapper(
        model.monitorMetricAlertDynamicCriteriaToTerraform,
        true,
      )(this.props.dynamicCriteria),
    });

    // Output properties
    this.id = metricAlert.id;
    const cdktfTerraformOutputMetricAlertId = new cdktf.TerraformOutput(
      this,
      "id",
      {
        value: metricAlert.id,
      },
    );
    cdktfTerraformOutputMetricAlertId.overrideLogicalId("id");
  }

  private ValidatePropsFrequency() {
    const frequencyOptions = ["PT1M", "PT5M", "PT15M", "PT30M", "PT1H"];
    if (!frequencyOptions.includes(this.props.frequency ?? "NotSet")) {
      throw new Error(`frequency must be one of ${frequencyOptions}`);
    }
  }

  private ValidatePropsWindowSize() {
    const windowSizeOptions = [
      "PT1M",
      "PT5M",
      "PT15M",
      "PT30M",
      "PT1H",
      "PT6H",
      "PT12H",
      "P1D",
    ];
    if (!windowSizeOptions.includes(this.props.windowSize ?? "NotSet")) {
      throw new Error(`windowSize must be one of ${windowSizeOptions}`);
    }
  }

  private ValidatePropsWindowSizeGreaterThanFrequency() {
    const f = moment.duration(this.props.frequency);
    const w = moment.duration(this.props.windowSize);
    if (w < f) {
      throw new Error("windowSize must be greater than frequency");
    }
  }
}
