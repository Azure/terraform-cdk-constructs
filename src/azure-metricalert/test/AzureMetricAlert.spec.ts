import { Testing, TerraformStack } from 'cdktf';
import 'cdktf/lib/testing/adapters/jest';
import { AzurermProvider } from "@cdktf/provider-azurerm/lib/provider";
import * as metricalert from "../../azure-metricalert";
import { exampleAzureMetricAlert } from './ExampleAzureMetricAlert';
import * as util from "../../util/azureTenantIdHelpers";


describe('Azure Metric Alert With Defaults', () => {
  let stack: TerraformStack;
  let fullSynthResult: any;

  beforeEach(() => {
    const app = Testing.app();
    stack = new TerraformStack(app, "test");

    new AzurermProvider(stack, "azureQueryRuleAlert", { features: {} });

    new metricalert.MetricAlert(stack, 'testAzureQueryRuleAlert', {
      name: "metric alert test",
      resourceGroupName: "testResourceGroup",
      scopes: ["/subscriptions/00000000-0000-0000-0000-000000000000"],
      criteria: [
        {
          metricName: "Heartbeat",
          metricNamespace: "Microsoft.operationalinsights/workspaces",
          aggregation: "Average",
          operator: "LessThan",
          threshold: 0,
        },
      ],
    });

    fullSynthResult = Testing.fullSynth(stack); // Save the result for reuse
  });

  it("renders an Azure Metric Alert with defaults and checks snapshot", () => {
    expect(
      Testing.synth(stack)
    ).toMatchSnapshot(); // Compare the already prepared stack
  });

  it("check if the produced terraform configuration is valid", () => {
    expect(fullSynthResult).toBeValidTerraform(); // Use the saved result
  });

  it("check if this can be planned", () => {
    expect(fullSynthResult).toPlanSuccessfully(); // Use the saved result
  });
});


describe('Linux Metric Alert Example', () => {

  it("renders the Metric Alert and checks snapshot", () => {
    // Need to remove the tenant_id from the snapshot as it will change wherever the test is run
    const output = Testing.synth(new exampleAzureMetricAlert(Testing.app(), "testAzureMetricAlert"));
    const myObject: Record<string, any> = JSON.parse(output);

    expect(util.removeTenantIdFromSnapshot(myObject)).toMatchSnapshot();
  });

  it("check if the produced terraform configuration is valid", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(Testing.fullSynth(new exampleAzureMetricAlert(Testing.app(), "testAzureMetricAlert"))).toBeValidTerraform();
  });

  it("check if this can be planned", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(Testing.fullSynth(new exampleAzureMetricAlert(Testing.app(), "testAzureMetricAlert"))).toPlanSuccessfully();
  });
});