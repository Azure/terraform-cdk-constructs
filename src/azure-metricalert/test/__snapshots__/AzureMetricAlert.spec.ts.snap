// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Azure Metric Alert With Defaults renders an Azure Metric Alert with defaults and checks snapshot 1`] = `
"{
  "output": {
    "id": {
      "value": "\${azurerm_monitor_metric_alert.testAzureQueryRuleAlert_metricAlert_7716E17C.id}"
    }
  },
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_monitor_metric_alert": {
      "testAzureQueryRuleAlert_metricAlert_7716E17C": {
        "auto_mitigate": true,
        "criteria": [
          {
            "aggregation": "Average",
            "metric_name": "Heartbeat",
            "metric_namespace": "Microsoft.operationalinsights/workspaces",
            "operator": "LessThan",
            "skip_metric_validation": false,
            "threshold": 0
          }
        ],
        "enabled": true,
        "frequency": "PT5M",
        "name": "metric alert test",
        "resource_group_name": "\${azurerm_resource_group.MyResourceGroup.name}",
        "scopes": [
          "/subscriptions/00000000-0000-0000-0000-000000000000"
        ],
        "severity": 3,
        "window_size": "PT5M"
      }
    },
    "azurerm_resource_group": {
      "MyResourceGroup": {
        "location": "eastus",
        "name": "rg-test"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.70.0"
      }
    }
  }
}"
`;
